### PR TITLE
Copter: attempt to improve setting home location while in indoor mode

### PR DESCRIPTION
**Pull request by iamczar**
As mentioned:

This is an attempt to set the home location sooner in case the phone is not available to set the home location.

### CODE REVIEW

File: ArduCopter/toy_mode.cpp [hunk 1]
```diff
@@ -416,6 +416,29 @@ void ToyMode::update()
...
+    
+    /*
+    * attemp to improve the setting of home location if it takes off when EKF is not happy
+    */
+    if(copter.motors->armed() && !copter.position_ok()){
+        // if we have a 3d lock and valid location
+        if(copter.gps.status() >= AP_GPS::GPS_OK_FIX_3D){
+            //aquire horizontal accuracy
+            float toy_horizontal_acc;
+            copter.gps.horizontal_accuracy(toy_horizontal_acc);
```
Comment (peterbarker) [inline ArduCopter/toy_mode.cpp:428]: Check for success; this is non-void

File: ArduCopter/toy_mode.cpp [hunk 2]
```diff
@@ -416,6 +416,29 @@ void ToyMode::update()
     } else {
         throttle_high_counter = 0;
     }
+    
+    /*
+    * attemp to improve the setting of home location if it takes off when EKF is not happy
+    */
+    if(copter.motors->armed() && !copter.position_ok()){
```
Comment (peterbarker) [inline ArduCopter/toy_mode.cpp:423]: nitpick: space between `)` and `{` (several more in this PR likewise)

File: ArduCopter/toy_mode.cpp [hunk 3]
```diff
@@ -416,6 +416,29 @@ void ToyMode::update()
     } else {
         throttle_high_counter = 0;
     }
+    
+    /*
+    * attemp to improve the setting of home location if it takes off when EKF is not happy
```
Comment (peterbarker) [inline ArduCopter/toy_mode.cpp:421]: spelling

File: ArduCopter/toy_mode.cpp [hunk 4]
```diff
@@ -416,6 +416,29 @@ void ToyMode::update()
...
+    */
+    if(copter.motors->armed() && !copter.position_ok()){
+        // if we have a 3d lock and valid location
+        if(copter.gps.status() >= AP_GPS::GPS_OK_FIX_3D){
+            //aquire horizontal accuracy
+            float toy_horizontal_acc;
+            copter.gps.horizontal_accuracy(toy_horizontal_acc);
+            //if gps horizontal accuracy of below 15m
+            if(toy_horizontal_acc <= 1500){
```
Comment (peterbarker) [inline ArduCopter/toy_mode.cpp:430]: Comment in the header says this is in metres....

File: ArduCopter/toy_mode.cpp [hunk 5]
```diff
@@ -416,6 +416,29 @@ void ToyMode::update()
...
+                //if home location has not been set yet
+                if(copter.ap.home_state == HOME_UNSET){
+                    //bypass EKF check and set currect location as HOME location
+                    Location indoor_loc = copter.gps.location(); 
+                    copter.ahrs.set_home(indoor_loc);
+                    copter.set_home_state(HOME_SET_NOT_LOCKED);
+                    GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "INDOOR MODE: setting home location here");
```
Comment (peterbarker) [inline ArduCopter/toy_mode.cpp:437]: Could replace "here" with some numbers :-)

**Comment by tridge** [comment]
apart from Peters comments, the main thing I'd like to see is some testing to indicate it does the right thing in the various circumstances. So when we get full GPS lock after takeoff, or we get better GPS lock while disarmed, or we get full GPS lock before takeoff.
In particular, update_home_from_EKF() checks for ap.home_state != HOME_UNSET. So it looks like this will disable the setting of home when the EKF gets lock. We definately want to use an improved home if we get it before we arm.

**Comment by iamczar** [comment]
Thanks for the comments, very useful. I'll do some flying today to cover the test scenarios

**Pull request closed without merging**